### PR TITLE
added "dump" to metadata command

### PR DESCRIPTION
**Pull request by afuechsel**
Also changed "remove" command to get it working even if the item is no longer existing. 

Signed-off-by: Andre Fuechsel <andre.fuechsel@telekom.de>

### CODE REVIEW

**Comment by afuechsel** [comment]
@htreu Please be aware, that this PR also contains a bug fix. So if the "dump" command will not be accepted, the bug fix is still needed.

**Comment by sjka** [comment]
I'm struggling hard to make up my mind. Somehow no strong opinion makes any attempt to form. 
In general I don't think it hurts, could be quite useful for development. But I don't see any potential use-case for operations.

The only thing I'm wondering is whether it really makes sense to give it a new name `dump`. Wouldn't it rather be more intuitive to change the make the two parameters of `list` optional like this?

```
smarthome metadata list [<itemName> [<namespace>]]
```

Then the following would work:

* `smarthome metadata list`
  would return _all_ metadata
* `smarthome metadata list foo`
  would return all metadata of item `foo`
* `smarthome metadata list foo bar`
  would return the `bar` metadata of `foo`

Another thing I don't really like is that we are exposing internal metadata here. IMHO that should be kept internal. The REST resource also doesn't allow access to them. It's just an implementation detail that the metadata infrastructure is used for e.g. tags. As is the format in which they are stored there. I don't consider this API it is subject to change.

**Comment by afuechsel** [comment]
Ok for list instead of dump. But namespace cannot be optional as there is no way to address metadata without a namespace. It’s part of MetadataKey. 

Exposing internal metadata - I don’t see an issue with that. As the console commands are mainly for developers, why not? May be with an optional parameter -internal? Inbox for example also allows to list hidden entries.

**Comment by sjka** [comment]
> But namespace cannot be optional as there is no way to address metadata without a namespace. It’s part of MetadataKey.

Correct. How about implementing it like this?

```java
metadataRegistry.stream().filter(MetadataPredicates.ofItem(itemName))...
```

**Comment by maggu2810** [comment]
> Another thing I don't really like is that we are exposing internal metadata here. IMHO that should be kept internal. The REST resource also doesn't allow access to them. It's just an implementation detail that the metadata infrastructure is used for e.g. tags.

Would it make sense to provide a function that returns only the non-internal ones?
If there are now at least two places that don't want to export the internal ones why should the caller filter it (and perhaps forget the filter)?

**Comment by afuechsel** [comment]
@maggu2810 As a developer I'd like to see everything, even implementation details. If you want I could add a `listinternal` command (like `listignored` for inbox) that lists the internal ones and list only to return the non-internal ones. Sometimes it's even helpful to see that "implementation detail", as it revealed the fact to me, that deleting the items did actually not delete their tags, so I could fix that bug (#5625).

**Comment by maggu2810** [comment]
@afuechsel My comment has not been about the console command. It is about about to make the usage of the API for Java Developers simpler and less error prune.
If every Java Developer that uses the ESH API needs to know that the metadata should be filtered to not export internal stuff to some external interface...
It would be more visible if there are two function (e.g. to get all metadata, to get only non-internal metadata) instead of rely that all developers knows that there are internal metadata and external one.

**Comment by afuechsel** [comment]
@maggu2810 I agree but I'd suggest that API changes should be handled in a different PR. I will update my PR to support `list` and `listinternal`.

File: bundles/io/org.eclipse.smarthome.io.console/src/main/java/org/eclipse/smarthome/io/console/internal/extension/MetadataConsoleCommandExtension.java [hunk 1]
```diff
@@ -91,21 +98,28 @@ public void execute(String[] args, Console console) {
...
+                Metadata metadata = metadataRegistry.get(key);
+                if (metadata != null) {
+                    console.println(metadata.toString());
+                }
             }
         }
     }
 
+    private boolean isInternal(@NonNull Metadata metadata, boolean internal) {
+        return metadataRegistry.isInternalNamespace(metadata.getUID().getNamespace()) == internal;
```
Comment (afuechsel) [inline bundles/io/org.eclipse.smarthome.io.console/src/main/java/org/eclipse/smarthome/io/console/internal/extension/MetadataConsoleCommandExtension.java]: While this is working - I'd suggest to extend the Metadata API in a way to have `MetadataRegistry#isInternalNamespace` to be static and having a method `isInternalNamespace` added to  `MetadataKey` and `Metadata` itself.

**Comment by afuechsel** [comment]
@maggu2810 More opinions regarding this PR?

File: bundles/io/org.eclipse.smarthome.io.console/src/main/java/org/eclipse/smarthome/io/console/internal/extension/MetadataConsoleCommandExtension.java [hunk 2]
```diff
@@ -87,14 +98,28 @@ public void execute(String[] args, Console console) {
...
+            if (metadataRegistry.isInternalNamespace(namespace) == internal) {
+                Metadata metadata = metadataRegistry.get(key);
+                if (metadata != null) {
+                    console.println(metadata.toString());
+                }
+            }
         }
     }
 
+    private boolean isInternal(@NonNull Metadata metadata, boolean internal) {
```
Comment (maggu2810) [inline bundles/io/org.eclipse.smarthome.io.console/src/main/java/org/eclipse/smarthome/io/console/internal/extension/MetadataConsoleCommandExtension.java]: We do not use `@NonNull` annotations instead we use `@NonNullByDefault` and `@Nullable`.

**Comment by maggu2810** [comment]
This PR does not contain the changes for adding a dump (list, list internal) command only but also fixes bugs.
Would be great if we could split the separate changes into different PRs next time.

Git provides us a great "revert" feature that does not work if we mix different stuff in a commit.
Also the history does not really fit as I do not expect a bugfix in a "adding feature" commit.

**Comment by afuechsel** [comment]
@maggu2810 you are right and I reverted the bugfix commit and created a separate PR: #5688

**Comment by maggu2810** [comment]
> @maggu2810 you are right and I reverted the bugfix commit and created a separate PR: #5688

Great! Thanks!

One additional for the remove bugfix? Then we merge all that three changes separate.

**Comment by afuechsel** [comment]
> One additional for the remove bugfix? Then we merge all that three changes separate.

?? Don't get it. The bugfix is removed from that PR and a separate PR created (where you already commented).

**Comment by maggu2810** [comment]
Hm, Github still shows me changes in `removedMetadata`

**Comment by afuechsel** [comment]
> Hm, Github still shows me changes in `removeMetadata` 

Sure, but as this PR is an extension of a newly introduced feature the change in `removeMetadata` is a bit related to the changes in the list command, where namespace is now optional, as @sjka suggested in on review comment. So I don't want to split this really small PR any further.

File: bundles/io/org.eclipse.smarthome.io.console/src/main/java/org/eclipse/smarthome/io/console/internal/extension/MetadataConsoleCommandExtension.java [hunk 3]
```diff
@@ -87,14 +95,29 @@ public void execute(String[] args, Console console) {
...
+            if (metadataRegistry.isInternalNamespace(namespace) == internal) {
+                Metadata metadata = metadataRegistry.get(key);
+                if (metadata != null) {
+                    console.println(metadata.toString());
+                }
+            }
         }
     }
 
+    @NonNullByDefault
```
Comment (maggu2810) [inline bundles/io/org.eclipse.smarthome.io.console/src/main/java/org/eclipse/smarthome/io/console/internal/extension/MetadataConsoleCommandExtension.java]: Missed it on review. We don't use `@NonNullByDefault` on methods but for the whole class.

Please have a look at our "conventions":
https://www.eclipse.org/smarthome/documentation/development/conventions.html#null-annotations

I will fix it in another PR.

**Pull request merged**